sycl/include/sycl/ext/oneapi/experimental/opportunistic_group.hpp

```diff
@@ -0,0 +1,148 @@
+#else
+    throw runtime_error("Non-uniform groups are not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+
+  id_type get_local_id() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return detail::CallerPositionInMask(Mask);
+#else
+    throw runtime_error("Non-uniform groups are not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+
+  range_type get_group_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return 1;
+#else
+    throw runtime_error("Non-uniform groups are not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+
+  range_type get_local_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return Mask.count();
+#else
+    throw runtime_error("Non-uniform groups are not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+
+  linear_id_type get_group_linear_id() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return static_cast<linear_id_type>(get_group_id()[0]);
+#else
+    throw runtime_error("Non-uniform groups are not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+
+  linear_id_type get_local_linear_id() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return static_cast<linear_id_type>(get_local_id()[0]);
+#else
+    throw runtime_error("Non-uniform groups are not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+
+  linear_id_type get_group_linear_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return static_cast<linear_id_type>(get_group_range()[0]);
+#else
+    throw runtime_error("Non-uniform groups are not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+
+  linear_id_type get_local_linear_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return static_cast<linear_id_type>(get_local_range()[0]);
+#else
+    throw runtime_error("Non-uniform groups are not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+
+  bool leader() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    uint32_t Lowest = static_cast<uint32_t>(Mask.find_low()[0]);
+    return __spirv_SubgroupLocalInvocationId() == Lowest;
+#else
+    throw runtime_error("Non-uniform groups are not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+
+private:
+  sub_group_mask Mask;
+
+protected:
+  opportunistic_group(sub_group_mask m) : Mask(m) {}
+
+  friend opportunistic_group this_kernel::get_opportunistic_group();
+};
+
+namespace this_kernel {
+
+inline opportunistic_group get_opportunistic_group() {
+#ifdef __SYCL_DEVICE_ONLY__
+#if defined(__SPIR__)
+  // TODO: It may be wiser to call the intrinsic than rely on this_group()
+  sycl::sub_group sg = sycl::ext::oneapi::this_sub_group();
+  sub_group_mask mask = sycl::ext::oneapi::group_ballot(sg, true);
+  return opportunistic_group(mask);
+#elif defined(__NVPTX__)
```

Comment:
DPC++ supports AMD GPU targets in addition to these targets.
Is it possible to give a diagnostic on unsupported targets?

Reply:
Honestly, I'm not sure.  I thought about this, but I was worried that if I emitted something like a `#warning` then it would trigger on every AMD GPU compilation that included the headers (rather than only compilations calling the function).

@steffenlarsen, is there a way to trigger a compile-time or run-time diagnostic when a function is invoked on an unsupported backend?  The only precedent I see for this is throwing a `runtime_error` on the host device.

Reply:
I agree. A warning would likely leak into any AMDGCN/NVPTX compilations. Had this function been template dependent we could have used a static assertion using the template argument. I believe something like

```c++
template <typename T = void, typename = std::enable_if_t<std::is_same_v<T, void>>>
inline opportunistic_group get_opportunistic_group() {
...
#elif defined(__NVPTX__) || defined (__AMDGCN__)
  static_assert(!std::is_same_v<T, void>, "get_opportunistic_group() is not currently supported on the target");
#endif
...
}
```
would work, but I am not sure it's worth the dependence.

Reply:
When we say "not currently supported" is this something that we expect to fix soon, or is it a long term problem?

I think the `static_assert` solution also does not work in the general case.  Imagine an application that has some kernels that run on Nvidia and some that run on Intel GPUs.  The application use `get_opportunistic_group` in a kernel that will be submitted to an Intel GPU and the host code checks the device type before submitting the kernel.  There are other kernels in the same TU that can run on Nvidia devices.

Wouldn't the `static_assert` trigger in that case also?

We invented the "optional kernel features" concept to handle cases like this.  The diagnostic occurs at runtime if the application mistakenly submits the kernel using `get_opportunistic_group` to an Nvidia device.

Reply:
> When we say "not currently supported" is this something that we expect to fix soon, or is it a long term problem?

From my perspective, this is a complete unknown right now, and it's why everything is marked experimental.

My aim is to get things running reliably across all the devices and backends, but we need to get implementation experience.  I sketched out `__SPIRV__` and `__NVPTX__` paths here because I understand what the implementations will look like; I need somebody with AMDGCN experience to take a look at things there.

Designing aspects and optional features for these group types right now seems premature to me, because we don't know which group types we'll need to make optional.  If it's not a lot of work for somebody to give us optional features for each group type then I'm not opposed to that, but I don't want to make a lot of extra work for folks if there's a chance we won't need the aspects later on.

Reply:
I wanted to make sure that when this experimental feature is used on unsupported platforms, users will receive a meaningful error message about it. Actually I think we might already have it. I see that only code under `#if defined(__SPIR__)` has `return` statement, so I expect if `__SPIR__` is not defined the compiler will report an error about missing `return`.
@Pennycook, do you have any tests for this API we can run on NVIDIA/AMD GPUs?

Reply:
The tests in https://github.com/intel/llvm-test-suite/pull/1574 will work on anything supporting a sub-group size of 32 (once the NVIDIA/AMD paths are implemented).  I've marked those backends as unsupported on NVIDIA/AMD for now because they're unimplemented, but @JackAKirk is working on implementing the missing parts.

Reply:
@JackAKirk, do you know whether DPC++ with #8202 gives any diagnostic on NVIDIA/AMD GPU for using new classes? I hope the compiler returns an error. Am I right?

John, meaningful diagnostic is nice to have, but should not block commit of experimental API. I suppose at this point all users of this feature know how to use it properly and what to do if it doesn't work as intended. We just need to make sure that when it's ready for widespread use, the behavior of the implementation is user-friendly.

Sorry for the inconvenience I cause by discussing non-critical matters.

Reply:
@gmlueck - I agree, the static assertion would cause problems in cases like the one you mentioned. Having an aspect for this and marking the relevant functions as requiring that aspect is definitely an option.

Reply:
For the groups like opportunistic_group that weren't implemented on NVPTX I think my memory is that the tests did not return an error immediately when `get_opportunistic_group` was called, but I think as @steffenlarsen points out this could maybe be fixed with a static assert like: https://github.com/intel/llvm/blob/9596ea0ddd362165cefd419e57cf2b57f5962c6c/sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp#L32
For all types added here except for tangle_group, I think that we resolved the discussions on what the NVPTX implementation should be, so I could push what I implemented for NVPTX after this PR is merged. For AMD I think that AMD gpus do not support independent forward progress and therefore don't have a similar instruction to __activemask from NVPTX, but I think that for AMD get_opportunistic_group can be supported in the essentially same was as it is for __SPIR__ here, by using ballot, the only thing is that amd also needs to support 64 WI subgroups.